lms/security.py

```diff
@@ -84,63 +94,62 @@ def forget(self, request):
+            # LTUser serialized as query param for authorization failures
+            "/api/d2l/oauth/authorize": query_param_policy,
+            "/api/blackboard/oauth/authorize": query_param_policy,
+            "/api/canvas_studio/oauth/authorize": query_param_policy,
+            "/api/canvas/oauth/authorize": query_param_policy,
             # To fetch pages content from LMSes' APIs
-            "/api/canvas/pages/proxy",
-            "/api/moodle/pages/proxy",
-        }:
-            return LTIUserSecurityPolicy(
-                partial(get_lti_user_from_bearer_token, location="querystring")
-            )
-
-        if path.startswith("/api") or path in {
-            "/lti/1.3/deep_linking/form_fields",
-            "/lti/1.1/deep_linking/form_fields",
-            "/lti/reconfigure",
-        }:
+            "/api/canvas/pages/proxy": query_param_policy,
+            "/api/moodle/pages/proxy": query_param_policy,
             # LTUser serialized in the headers for API calls from the frontend
-            return LTIUserSecurityPolicy(
-                partial(get_lti_user_from_bearer_token, location="headers")
-            )
-
-        if path in {"/assignment", "/assignment/edit"} or path.startswith(
-            "/dashboard/launch/assignment/"
-        ):
+            re.compile(r"^/api.*"): headers_policy,
+            "/lti/1.3/deep_linking/form_fields": headers_policy,
+            "/lti/1.1/deep_linking/form_fields": headers_policy,
+            "/lti/reconfigure": headers_policy,
             # LTUser serialized in a from for non deep-linked assignment configuration
-            return LTIUserSecurityPolicy(
-                partial(get_lti_user_from_bearer_token, location="form")
-            )
+            "/assignment": form_policy,
+            "/assignment/edit": form_policy,
+            re.compile(r"^/dashboard/launch/assignment/\d+"): form_policy,
+            # LTIUser serialized in a cookie, for dashboards
+            re.compile(r"^/dashboard/assignment/\d+"): cookie_policy,
+        }
 
-        if path.startswith("/dashboard/assignment/"):
-            return LTIUserSecurityPolicy(
-                partial(get_lti_user_from_bearer_token, location="cookies")
-            )
+    @staticmethod
+    def get_policy(request: Request):
+        """Pick the right policy based the request's path."""
+
+        for pattern, policy in SecurityPolicy.path_policy_map().items():
+            match pattern:
+                case str():
+                    if pattern == request.path:
+                        return policy()
 
-        if path in {"/email/preferences", "/email/unsubscribe"}:
+                case re.Pattern():
+                    if pattern.match(request.path):
+                        return policy()
+
+                case _:  # pragma: no cover
+                    raise ValueError()
+
+        if request.path in {"/email/preferences", "/email/unsubscribe"}:
```

Comment:
This is a bit unsatisfactory, still having this outside the patter map.

If we want to refactor it I'd prefer to do it in a separate PR so this one doesn't keep growing.